File: 2784-check-if-array-is-good/2784-check-if-array-is-good.rb
# @param {Integer[]} nums
# @return {Boolean}
def is_good(nums)
   n = nums.size
   nums.sort!
   # sort nums and check
   # 1. every number less than n appears once
   # 2. n - 1 appears twice
   for i in (1..n-2) do 
      if i != nums[i-1]
         return false
      end
   end
   
   # if the last two 
   nums[-1] == n-1 && nums[-2] == n-1
end

File: 2784-check-if-array-is-good/2784-check-if-array-is-good.ts
function isGood(nums: number[]): boolean {
    // every number appears exactly once except the largest
    const mx = nums.length - 1 // this should be the largest num
    let xor = 0
    let mxN = 0
    for (let i = 0; i < nums.length - 1; xor ^= i++);
    
    for (const num of nums) {
        if (num > mx) {
            return false 
        } else if (num == mx) {
            mxN++
        } else {
            xor ^= num
        }
    }
    return xor == 0 && mxN == 2
};